storage/inmem/inmem.go

```diff
@@ -37,16 +38,22 @@ func New() storage.Store {
 // NewWithOpts returns an empty in-memory store, with extra options passed.
 func NewWithOpts(opts ...Opt) storage.Store {
 	s := &store{
-		data:             map[string]interface{}{},
-		triggers:         map[*handle]storage.TriggerConfig{},
-		policies:         map[string][]byte{},
-		roundTripOnWrite: true,
+		triggers:              map[*handle]storage.TriggerConfig{},
+		policies:              map[string][]byte{},
+		roundTripOnWrite:      true,
+		returnASTValuesOnRead: false,
 	}
 
 	for _, opt := range opts {
 		opt(s)
 	}
 
+	if s.returnASTValuesOnRead {
+		s.data = ast.NewObject()
+	} else {
+		s.data = map[string]interface{}{}
+	}
+
```

Comment:
Interesting..we are controlling how data is written to the store based on `returnASTValuesOnRead`. Did you consider always writing AST values to the store and return AST/Go values based on `returnASTValuesOnRead`.

Reply:
Always converting data to AST values would have a huge impact on OPA:s memory footprint and on bundle update performance. If these weren't concerns, then I think there is an argument for completely dropping the old raw data option altogether, as `topdown` is agnostic to the data type. Unfortunately, I think this feature is something only for users that either have very little data (at which point the performance boost is likely pretty small anyways), or have the ability to sacrifice these things to improve eval performance.

Reply:
Having an option to do lazy conversion, but storing converted values for re-use (which I've mentioned in other places), might be a middle ground that might be enough of the best of both worlds that it could be enabled by default. This is however an unsubstantiated claim that I haven't thought too hard about.